### PR TITLE
Award new users with V2X Tokens on ethereum

### PR DESCRIPTION
**Pull request by shumail92**
This PR includes couple of things:
- Ethereum service: Generic service for interacting with ethereum blockckahin.
- Notifications service: Generic service for sending email notifications
- SignupController: Signup now triggers `sendSignupBonus` that sends newly registered user V2X Tokens and on successful transaction, it sends email.
- For emails, we use Mailgun Transactional email service. They provide 10,000 emails/month for free. I guess that's enough for now. We'll upgrade or switch if we need more in future.

VE-104

### CODE REVIEW

File: src/services/ethereum.service.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    // check if wallet address is valid
+    if (!this.web3.utils.isAddress(wallet)) {
+      this.logger.error(config.errors.ethereum.invalidAddress, wallet);
+      throw Error(config.errors.ethereum.invalidAddress + ' ' + wallet);
+    }
+    try {
+      // get contract instance
+      let tokenContract = new this.web3.eth.Contract(this.contractAbi, this.contractAddress);
+      // get gas price & use medium value
+      let gasPriceGwei = (await this.getCurrentGasPrices()).medium;
```
Comment (v2xsami) [inline src/services/ethereum.service.ts]: What the heck is "gas price" ? I don't think it means the price of gasoline at the tankstelle, right?

File: src/services/ethereum.service.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      // set gas limit
+      let gasLimit = 100000;
+      // get tx count. we use it as nounce for new transaction
```
Comment (v2xsami) [inline src/services/ethereum.service.ts]: typo: nounce/nonce

File: src/services/notifications.service.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    });
+  }
+
+  /**
+   * send email notification
+   * @param {Mailgun.messages.SendData} emailObject Attributes of email to be sent.
+   */
+  sendEmailNotification(emailObject: Mailgun.messages.SendData): Promise<Boolean> {
+    const emailBody = emailObject.html;
+    const emailFooter = '<br/><br/>copyright &copy; 2018 V2XNetwork';
```
Comment (v2xsami) [inline src/services/notifications.service.ts]: adding the footer in this function doesn't seem like the right place to do it

File: src/controllers/signupController.ts [hunk 4]
```diff
@@ -51,3 +59,46 @@ signupRouter.route('/')
...
+        to: user.email,
+        subject: 'V2X Network - Signup bonus sent!',
+        html: `<p>
+        Dear ${user.firstName} ${user.lastName},
+        <br/><br/>
+        Thank you for signing up at V2X Network. We have sent a bonus of 1,000 V2X Tokens to your wallet: <b>${receipt.to}</b>.
+        <br /><br />
+        Here's the Etherscan link to transaction: <a href="${etherscanUrl}/tx/${receipt.transactionHash}" target="_blank">Etherscan Link.</a>
+        <br /><br />
+        If you have any questions, feel free to contact us.
```
Comment (v2xsami) [inline src/controllers/signupController.ts]: Add contact details here, for example email sendto link

Comment (v2xsami) [reply src/controllers/signupController.ts]: Add the email footer here (if needed) or perhaps make another function such as "formatV2XEmail(...)" which would append the footer.

File: src/services/ethereum.service.ts (hunk 1 above)
Comment (shumail92) [reply src/services/ethereum.service.ts]: haha no. It's the Gas Price for Ethereum network. The costs for computations on Ethereum network are calculated in terms of `gas`. And each `gas` unit is charged in `ether`. And this is what I am doing here - Setting the price according to current recommendations that i fetch from https://ethgasstation.info

**Comment by shumail92** [comment]
@v2xsami Regarding the fault tolerance in architecture, you are correct that there's no retry or any other fault tolerance at the moment - which i don't like it either. I am not sure if storing the states in database is a nice approach here and retrying on the basis of that. Rather, I think that we should decouple these background services from core in this case. For example, by introducing queues in between where core throws a ethereum-transaction job on the queue and then some worker picks the job from queue, mark it as "in process", processes the job and on success, sends acknowledgement to queue and removes the job from queue. In case the worker goes down, another worker can take the same job once the timeout happens and status of job changes from 'in-process' to 'todo'.
I quickly googled for queue options on Google cloud and they have Google Cloud Pub/Sub messaging service. Never used it, so will look into it or other possibilities.


Regarding the Ethereum address, I thought about it but then I guess it's fine to have the ethereum address as mandatory because that's a way of promoting blockchain adoption.
In the Sign-up form, I can add a link to https://myetherwallet.com so the users who don't have ethereum wallet, can just create one quickly. And then with our signup bonus transaction, their wallets won't be empty anymore. 
So I'll prefer to keep this requirement as mandatory.

@a-shamim Do you have any thing to add here or any thoughts on this mandatory requirement?

**Comment by v2xsami** [comment]
@shumail92 yes, correct they should be separate services. But still I would not store the state of the account in the queue, rather in the database. message queue would only be used to notify some service to do some work. 

We shouldn't want to promote ethereum per se. We should promote v2x network software. IMHO any step that makes it more complicated to sign up should be removed.

**Comment by a-shamim** [comment]
i.d say for now lets keep having the wallet mandatory, but we should try to make it super easy somehow for the user to create one maybe by providing a link directly to one of the easiest wallet creation site.

**Comment by a-shamim** [comment]
Not to forget its a POC, we can always revise this strategy based on input from actual users

File: src/controllers/signupController.ts (hunk 4 above)
Comment (shumail92) [reply src/controllers/signupController.ts]: I'll skip footer for now.

File: src/controllers/signupController.ts [hunk 5]
```diff
@@ -51,3 +63,64 @@ signupRouter.route('/')
...
+ */
+const sendSignupBonus = async (user: User) => {
+  // avoid sending any signup bonus during test. I know this is shitty but can't think of any elegant solution for now.
+  if (process.env.NODE_ENV !== 'test') {
+    logger.debug(`Attempting to send ${NUM_Of_BONUS_TOKENS} V2X Tokens as signup bonus to user: `, JSON.stringify(user));
+    const ethereumService = new EthereumService();
+    try {
+      // todo: check for one signup bonus <=> 1 eth address
+      // transfer tokens
+      let receipt = await ethereumService.transferTokens(user.ethWalletAddress, NUM_Of_BONUS_TOKENS);
```
Comment (v2xsami) [inline src/controllers/signupController.ts]: The problem here is that I don't like the idea of this API call waiting some other API call like that ethereum service. How long will it take to complete? Can *this* API call timeout? 

Secondary problem is that if for example that ehtereumService.transferTokens throw an exception the execution jumps into exception handler and doesn't record the transaction in the database. 

Finally is the user creation + sign up bonus INSERT atomic in a DB transaction?

**Pull request merged**